[Model Averaging] Create a base class for model averaging

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62132
* #62131
* __->__ #62111

This base class will be passed to the post-localSGD optimizer in the next PR. This way, the same post-localSGD optimizer can choose different model averaging algorithms.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29884954](https://our.internmc.facebook.com/intern/diff/D29884954/)

Commit:
[Model Averaging] Create a base class for model averaging

This base class will be passed to the post-localSGD optimizer in the next PR. This way, the same post-localSGD optimizer can choose different model averaging algorithms.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29884954](https://our.internmc.facebook.com/intern/diff/D29884954/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/model_averaging/averagers.py:
```diff
@@ -4,7 +4,29 @@
+        )
+        self.step = 0
+
+    def average_parameters(self):
```
Should we consider making `ModelAverager` an abstract class and enforcing that `average_parameters` is overridden, or at least raise a NotImplementedError here?

rohan-varma reviewed (approved):
lgtm, one nit

rohan-varma commented on torch/distributed/algorithms/model_averaging/averagers.py:
```diff
@@ -4,7 +4,29 @@
+        pass
+
+
+class PeriodicModelAverager(ModelAverager):
```
Do we plan to add additional model averaging impls in the future? If so, do we have an idea what those could be?

SciPioneer replied:
I do have prototyped another more complicated implementation, and SlowMo can be another candidate here. The point is that when I wrap this kind of model averaging component in an optimizer, I want to use an interface as the typing info.

Commit:
Update on "[Model Averaging] Create a base class for model averaging"


This base class will be passed to the post-localSGD optimizer in the next PR. This way, the same post-localSGD optimizer can choose different model averaging algorithms.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29884954](https://our.internmc.facebook.com/intern/diff/D29884954/)

[ghstack-poisoned]
